src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp

```diff
@@ -42,47 +42,60 @@ inline T UnifiedOopRef::addr() const {
 // this specialization provides a workaround.
 template<>
 inline uintptr_t UnifiedOopRef::addr<uintptr_t>() const {
-  return _value & ~uintptr_t(3);
+  return (_value & ~tag_mask) LP64_ONLY(>> 1);
 }
 
 inline bool UnifiedOopRef::is_narrow() const {
-  return _value & 1;
+  return (_value & narrow_tag) != 0;
 }
 
 inline bool UnifiedOopRef::is_native() const {
-  return _value & 2;
+  return (_value & native_tag) != 0;
+}
+
+inline bool UnifiedOopRef::is_raw() const {
+  return (_value & raw_tag) != 0;
 }
 
 inline bool UnifiedOopRef::is_null() const {
   return _value == 0;
 }
 
-inline UnifiedOopRef UnifiedOopRef::encode_in_native(const narrowOop* ref) {
+template <typename T>
+inline UnifiedOopRef create_with_tag(T ref, uintptr_t tag) {
   assert(ref != NULL, "invariant");
-  UnifiedOopRef result = { reinterpret_cast<uintptr_t>(ref) | 3 };
-  assert(result.addr<narrowOop*>() == ref, "sanity");
+  assert(((reinterpret_cast<uintptr_t>(ref) LP64_ONLY(<< 1)) & UnifiedOopRef::tag_mask) == 0, "Unexpected low-order bits");
+  LP64_ONLY(assert(((reinterpret_cast<uintptr_t>(ref)) & (1ull << 63)) == 0, "Unexpected high-order bit"));
+  UnifiedOopRef result = { (reinterpret_cast<uintptr_t>(ref) LP64_ONLY(<< 1)) | tag };
+  assert(result.addr<T>() == ref, "sanity");
```

Comment:
This is crazy hard to read.

Reply:
Yeah. Refactoring out `reinterpret_cast<uintptr_t>(ref)` into a variable might help a bit at least. The assert just want to check that the lowest two bits are not set (the tagging does not destroy information) and that on 64 bit platforms the highest bit is not set (the shift does not destroy information).

Not sure if there is a better way to express those two invariants.

Reply:
Maybe it would be easier to read the code if it was rewritten as (uncompiled):
```
  uintptr_t value = reinterpret_cast<uintptr_t>(ref);

#ifdef _LP64
  // tag_mask is 3 bits. When ref is a narrowOop* we only have 2 alignment
  // bits, because of the 4 byte alignment of compressed oops addresses.
  // Shift up to make way for one more bit.
  assert(value & (1ull << 63)) == 0, "Unexpected high-order bit"));
  value <<= 1;
#endif

  UnifiedOopRef result = { value | tag };
  assert(result.addr<T>() == ref, "sanity");
```